napari/layers/labels/_labels_mouse_bindings.py

```diff
@@ -35,48 +29,20 @@ def draw(layer, event):
     else:
         new_label = layer.selected_label
 
-    if coordinates is not None:
-        if layer._mode in [Mode.PAINT, Mode.ERASE]:
-            layer.paint(coordinates, new_label)
-        elif layer._mode == Mode.FILL:
-            layer.fill(coordinates, new_label)
-    else:  # still add an item to undo queue
-        # when dragging, if we start a drag outside the layer, we will
-        # incorrectly append to the previous history item. We create a
-        # dummy history item to prevent this.
-        dummy_indices = (np.zeros(shape=0, dtype=int),) * layer.data.ndim
-        layer._undo_history.append([(dummy_indices, [], [])])
-
-    last_cursor_coord = coordinates
+    layer._draw(new_label, coordinates, coordinates)
     yield
 
-    layer._block_saving = True
+    last_cursor_coord = coordinates
     # on move
     while event.type == 'mouse_move':
         coordinates = mouse_event_to_labels_coordinate(layer, event)
         if coordinates is not None or last_cursor_coord is not None:
-            interp_coord = interpolate_coordinates(
-                last_cursor_coord, coordinates, layer.brush_size
-            )
-            for c in interp_coord:
-                if (
-                    ndisplay == 3
-                    and layer.data[tuple(np.round(c).astype(int))] == 0
-                ):
-                    continue
-                if layer._mode in [Mode.PAINT, Mode.ERASE]:
-                    layer.paint(c, new_label, refresh=False)
-                elif layer._mode == Mode.FILL:
-                    layer.fill(c, new_label, refresh=False)
-            layer.refresh()
+            layer._draw(new_label, last_cursor_coord, coordinates)
         last_cursor_coord = coordinates
         yield
 
     # on release
-    layer._block_saving = False
-    undo_item = layer._undo_history[-1]
-    if len(undo_item) == 1 and len(undo_item[0][0][0]) == 0:
-        layer._undo_history.pop()
+    layer._finish_painting()
```

Comment:
@DragaDoncila so the thing that I always wanted to do here but never quite managed was something like:

```python
with layer.block_history():
    <painting_loop>
```

Then in the layer you would have:

```python
@contextmanager
def block_history(self):
    self._block_history = True
    yield
    self._block_history = False
```

Finally, *all* of paint, fill etc. would do:

```python
self.save_history(history_atom)
```

Finally, `self.save_history` would be the one function that would check `self._block_history`, and either append to the staging item or append the staging item to the history.

The advantages of this approach are:
- Use a context manager for something that smells very much like a context
- rather than move the append/add to history one level down to the paint and fill methods, move it all the way where it belongs in the save_history method

What do you think?

Reply:
@jni 
when using `contextmanager` are used an exception is raised inside `with` clause in `yield` place. As your example does not contain a handling exception then stat could be blocked. 

See that in the documentation there is the suggestion to use `try: ... finally:`
https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager

It should looks like:

```
@contextmanager
def block_history(self):
    prev = self._block_history
    self._block_history = True
    try:
        yield
    finally: 
        self._block_history = prev
```

Reply:
Thanks for that snippet @Czaki! 🙏